### PR TITLE
Create a single file operator-install.yaml bundle to install the operator

### LINKED ISSUE #445: Create a single file install.yaml file to install the operator
It could be easier to setup the operator in kubernetes if we can just do `kubectl apply -f singlefile.yaml` to install the operator. So I suggest to put the different files in one single bundle.

### PR DESCRIPTION
**Pull request by dmvolod**
Fixes #445

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -68,22 +68,29 @@ $ export KUBECONFIG=/path/to/admin.kubeconfig
   ```
   $ make run-local
   ```
-4. Open a new terminal window and create an Infinispan cluster with three nodes:
+4. Install the Infinispan operator on OKD from the bundle file (option).
+  * Create a new project or use existing one:
```
Comment (oraNod) [inline README.md:72]:
```suggestion
  * Create a new project or use an existing one:
```

File: README.md [hunk 2]
```diff
@@ -68,22 +68,29 @@ $ export KUBECONFIG=/path/to/admin.kubeconfig
   ```
   $ make run-local
   ```
-4. Open a new terminal window and create an Infinispan cluster with three nodes:
+4. Install the Infinispan operator on OKD from the bundle file (option).
```
Comment (oraNod) [inline README.md:71]:
```suggestion
4. Optionally install the Infinispan operator on OKD from the bundle file.
```

File: README.md [hunk 3]
```diff
@@ -68,22 +68,29 @@ $ export KUBECONFIG=/path/to/admin.kubeconfig
...
+  ```
+  $ oc apply -n infinispan-operator -f https://raw.githubusercontent.com/infinispan/infinispan-operator/master/deploy/operator-install.yaml
+  ```
+5. Open a new terminal window and create an Infinispan cluster with two nodes:
 ```
 $ oc apply -f deploy/cr/minimal/cr_minimal.yaml
 infinispan.infinispan.org/example-infinispan configured
 ```
-5. Watch the pods start until they start running.
+6. Watch the pods start until they start running.
```
Comment (oraNod) [inline README.md:85]:
```suggestion
6. Watch the pods until they are running.
```

File: README.md [hunk 4]
```diff
@@ -96,21 +103,29 @@ $ make minikube-config
...
+  * Run this command to install the operator:
+  ```
+  $ kubectl apply -n local-operators -f https://raw.githubusercontent.com/infinispan/infinispan-operator/master/deploy/operator-install.yaml
+  ```
+5. Open a new terminal window and create an Infinispan cluster with two nodes:
 ```bash
 $ kubectl apply -f deploy/cr/minimal/cr_minimal.yaml -n local-operators
 ```
-5. Watch the pods start until the start running
+6. Watch the pods start until the start running
```
Comment (oraNod) [inline README.md:123]:
```suggestion
6. Watch the pods until they are running
```

**Comment by dmvolod** [comment]
Thanks, @oraNod . Fixed.

File: README.md [hunk 5]
```diff
@@ -68,22 +68,29 @@ $ export KUBECONFIG=/path/to/admin.kubeconfig
...
   ```
-4. Open a new terminal window and create an Infinispan cluster with three nodes:
+4. Optionally install the Infinispan operator on OKD from the bundle file.
+  * Create a new project or use an existing one:
+  ```
+  $ oc new-project infinispan-operator
+  ```
+  * Run this command to install the operator:
+  ```
+  $ oc apply -n infinispan-operator -f https://raw.githubusercontent.com/infinispan/infinispan-operator/master/deploy/operator-install.yaml
```
Comment (rigazilla) [inline README.md:78]: I would use the same format for all the `oc apply` commands: below we're using local file and no project specified

File: README.md [hunk 6]
```diff
@@ -96,21 +103,29 @@ $ make minikube-config
...
 ```
-4. Open a new terminal window and create an Infinispan cluster with three nodes:
+4. Optionally install the Infinispan operator on Kubernetes from the bundle file.
+  * Create a new namespace or use an existing one:
+  ```
+  $ kubectl create namespace local-operators
+  ```
+  * Run this command to install the operator:
+  ```
+  $ kubectl apply -n local-operators -f https://raw.githubusercontent.com/infinispan/infinispan-operator/master/deploy/operator-install.yaml
```
Comment (rigazilla) [inline README.md:117]: see comment above.
Could we use also here `infinispan-operator` as namespace?

File: build/install-bundle.sh [hunk 7]
```diff
@@ -0,0 +1,29 @@
+#!/usr/bin/env bash
+
+OPERATOR_SOURCE_INSTALL_FILES=(
+  "deploy/role.yaml"
+  "deploy/service_account.yaml"
+  "deploy/role_binding.yaml"
+  "deploy/crds/infinispan.org_infinispans_crd.yaml"
```
Comment (rigazilla) [inline build/install-bundle.sh:7]: we probably want also `clusterrole.yaml` and `clusterrole_binding.yaml` needed for the cross site feature?

Comment (dmvolod) [reply build/install-bundle.sh:7]: Yes, sure.
But got following error during install:
**The ClusterRoleBinding "infinispan-operator" is invalid: subjects[0].namespace: Required value**

This means that ClusterRoleBinding need to be defined namespace for account. This is limiting for the flexible install in any namespace/project

**Comment by dmvolod** [comment]
@rigazilla fixed, but still awaiting your feedback related to ClusterRoleBinding account namespace

File: build/install-bundle.sh (hunk 7 above)
Comment (rigazilla) [reply build/install-bundle.sh:9]: It seems it's not possible to imply the current namespace in the subject. Really annoying!
I can see only 2 options:

- fix a namespace (at least for the serviceaccount)
- split the install in two steps

they are equally bad to me :-)

Comment (dmvolod) [reply build/install-bundle.sh:9]: Just another idea, tag this namespace with something specific, like $infinispan-operator-ns and replace with sed as I did in push-okd4.sh

Comment (rigazilla) [reply build/install-bundle.sh:9]: What do you think about:
- implement your approach in the `make install-bundle` to accept the custom namespace and sed all the .yaml for the users that need to specify namespace
- leave the yaml with fixed `infinispan-operator` project: this for users that want to quickstart with standard oc/kubectl commands

this to keep the example clean with only oc/kubectl commands for people that want to quickstart

Comment (dmvolod) [reply build/install-bundle.sh:9]: @rigazilla I'm prefer much simple use case as possible for this issue. Let's keep `make install-bundle` for joining files content only and add notification, that customer needs to replace this namespace with your own one. Strimzi project do this
https://strimzi.io/docs/operators/master/deploying.html#deploying-cluster-operator-str
wdyt?

File: deploy/operator-install.yaml [hunk 8]
```diff
@@ -0,0 +1,742 @@
...
+  - customresourcedefinitions
+  - customresourcedefinitions/status
+  verbs:
+  - get
+  - list
+  - watch
+- apiGroups:
+  - ""
+  resources:
+  - nodes/metrics
```
Comment (dmvolod) [inline deploy/operator-install.yaml:111]: @rigazilla can we join this resources with upper `nodes` section? Do we really need `watch` verb? I didn't see any watch subscriptions.

File: deploy/operator-install.yaml [hunk 9]
```diff
@@ -0,0 +1,742 @@
...
+    name: infinispan-operator
+roleRef:
+  kind: Role
+  name: infinispan-operator
+  apiGroup: rbac.authorization.k8s.io
+
+---
+
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRoleBinding
```
Comment (dmvolod) [inline deploy/operator-install.yaml:139]: @rigazilla I would prefer to put ClusterRoleBinding as top as possible and comment that customer need to change a namespace for deployment outside the defined namespace

File: build/install-bundle.sh (hunk 7 above)
Comment (rigazilla) [reply build/install-bundle.sh:9]: @dmvolod ok simple enough: let's leave `infinispan-operator` and add a note aside for users that want to change namespace.

**Comment by dmvolod** [comment]
@rigazilla also `make run-local` doesn't utilize ClusterRole* resources, but I'm not sure that we should need to add xsite support for local running on dev stage. Or we should also change?

**Comment by rigazilla** [comment]
> @rigazilla also `make run-local` doesn't utilize ClusterRole* resources, but I'm not sure that we should need to add xsite support for local running on dev stage. Or we should also change?

I think we can skip this, when developing I usually run the operator with my `oc login` account, not with the service-account. Wdyt?

**Comment by dmvolod** [comment]
> > @rigazilla also `make run-local` doesn't utilize ClusterRole* resources, but I'm not sure that we should need to add xsite support for local running on dev stage. Or we should also change?
> 
> I think we can skip this, when developing I usually run the operator with my `oc login` account, not with the service-account. Wdyt?

Yes, sure, but in this case we should looks like a little bit validate and cleanup `run-local` steps

**Comment by rigazilla** [comment]
> Yes, sure, but in this case we should looks like a little bit validate and cleanup `run-local` steps

Right, I think we could have a separated issue for this assessment.

**Comment by dmvolod** [comment]
> > Yes, sure, but in this case we should looks like a little bit validate and cleanup `run-local` steps
> 
> Right, I think we could have a separated issue for this assessment.

@rigazilla, created a new issue #476 and fixed docs with `sed` command. Please review again, thanks.

File: deploy/operator-install.yaml (hunk 8 above)
Comment (dmvolod) [reply deploy/operator-install.yaml:111]: @rigazilla could you please to look at the ClusterRole to optimize some permissions and compact file as small as possible.

Comment (rigazilla) [reply deploy/operator-install.yaml]: @dmvolod, I've merged the `nodes` section and removed watch

Comment (dmvolod) [reply deploy/operator-install.yaml]: Thanks, @rigazilla

File: README.md [hunk 10]
```diff
@@ -68,22 +68,36 @@ $ export KUBECONFIG=/path/to/admin.kubeconfig
   ```
   $ make run-local
   ```
-4. Open a new terminal window and create an Infinispan cluster with three nodes:
+4. Optionally install the Infinispan operator on OKD from the bundle file.
+  * Create a new project or use an existing one:
+  ```
+  $ oc new-project infinispan-operator
+  ```
+  * If your project name different from the `infinispan-operator` (in this example `my-infinispan-operator`), download operator install bundle to the local store and replace namespace with the following command:
```
Comment (rigazilla) [inline README.md:76]:
```suggestion
  * If your project name differs from `infinispan-operator` (in this example `my-infinispan-operator`), you need to download the operator install bundle file to the local store and replace the predefined namespace with the following command:
```

Comment (rigazilla) [reply README.md:76]: If you're supposing that the user doesn't checkout the whole project, you probably need to provide info on how to get the `cr_minimal.yaml` below in point 5.

Comment (rigazilla) [reply README.md:76]: maybe @oraNod can provide a better rephrase for the param above.

File: README.md [hunk 11]
```diff
@@ -96,21 +110,36 @@ $ make minikube-config
...
-3. Build the operator and run it locally:
+3. Build the operator from the source code and run it locally:
 ```bash
 $ make minikube-run-local
 ```
-4. Open a new terminal window and create an Infinispan cluster with three nodes:
+4. Optionally install the Infinispan operator on Kubernetes from the bundle file.
+  * Create a new namespace or use an existing one:
+  ```
+  $ kubectl create namespace infinispan-operator
```
Comment (rigazilla) [inline README.md:120]: see comments for the openshift section

File: README.md (hunk 10 above)
Comment (dmvolod) [reply README.md:76]: > If you're supposing that the user doesn't checkout the whole project, you probably need to provide info on how to get the `cr_minimal.yaml` below in point 5.

@rigazilla let's point to the raw.githubusercontent.com whether local or remote creating

Comment (rigazilla) [reply README.md]: ok

**Comment by dmvolod** [comment]
> lgtm, please squash the commits

Done, @rigazilla 
Thanks for review

File: README.md [hunk 12]
```diff
@@ -68,22 +68,36 @@ $ export KUBECONFIG=/path/to/admin.kubeconfig
   ```
   $ make run-local
   ```
-4. Open a new terminal window and create an Infinispan cluster with three nodes:
+4. Optionally install the Infinispan operator on OKD from the bundle file.
+  * Create a new project or use an existing one:
+  ```
+  $ oc new-project infinispan-operator
+  ```
+  * If your project name differs from `infinispan-operator` (in this example `my-infinispan-operator`), you need to download the operator install bundle file locally and replace the predefined namespace with the following command:
```
Comment (oraNod) [inline README.md:76]:
```suggestion
  * If your project name is not `infinispan-operator`, download the Operator install bundle file and replace the pre-defined namespace. For example, the following command changes the project name to `my-infinispan-operator`:
```

Comment (dmvolod) [reply README.md]: Thanks, @oraNod , fixed

**Comment by dmvolod** [comment]
Thanks ,@rigazilla and @oraNod 
Merged

**Pull request merged**